kernel/src/engine/arrow_utils.rs

```diff
@@ -1,89 +1,994 @@
+                                input_fields[parquet_position].is_nullable(),
+                            ));
+                            final_fields_cols[*index] = Some((new_field, Arc::new(result_array)));
+                        }
+                        ArrowDataType::List(_) => {
+                            let list_array = input_cols[parquet_position].as_list::<i32>().clone();
+                            final_fields_cols[*index] = reorder_list(
+                                list_array,
+                                input_fields[parquet_position].name(),
+                                children,
+                            )?;
+                        }
+                        ArrowDataType::LargeList(_) => {
+                            let list_array = input_cols[parquet_position].as_list::<i64>().clone();
+                            final_fields_cols[*index] = reorder_list(
+                                list_array,
+                                input_fields[parquet_position].name(),
+                                children,
+                            )?;
+                        }
+                        // TODO: MAP
+                        _ => {
+                            return Err(Error::generic(
+                            "Child reorder can only apply to struct/list/map. This is a kernel bug, please report"
+                        ));
+                        }
+                    }
+                }
+                ReorderIndex::Index { index } => {
+                    final_fields_cols[*index] = Some((
+                        input_fields[parquet_position].clone(), // cheap Arc clone
+                        input_cols[parquet_position].clone(),   // cheap Arc clone
+                    ));
+                }
+                ReorderIndex::Null { index, field } => {
+                    let null_arry = Arc::new(new_null_array(field.data_type(), num_rows));
+                    let field = field.clone(); // cheap Arc clone
+                    final_fields_cols[*index] = Some((field, null_arry));
+                }
+            }
+        }
+        let mut field_vec = Vec::with_capacity(num_cols);
+        let mut reordered_columns = Vec::with_capacity(num_cols);
+        for field_array_opt in final_fields_cols.into_iter() {
+            let (field, array) = field_array_opt.ok_or_else(|| {
+                Error::generic(
+                    "Found a None in final_fields_cols. This is a kernel bug, please report.",
+                )
+            })?;
+            field_vec.push(field);
+            reordered_columns.push(array);
+        }
+        Ok(StructArray::try_new(
+            field_vec.into(),
+            reordered_columns,
+            null_buffer,
+        )?)
+    }
+}
+
+fn reorder_list<O: OffsetSizeTrait>(
+    list_array: GenericListArray<O>,
+    input_field_name: &str,
+    children: &[ReorderIndex],
+) -> DeltaResult<FieldArrayOpt> {
```

Comment:
It lets me reuse the type alias. I think this looks nicer than `DeltaResult<(Arc<ArrowField>, Arc<dyn ArrowArray>)`, and if this didn't return an option, I'd just have to wrap it in one at the call site since the result is put into `final_fields_cols` which is holding options.

Reply:
Gotcha. Tho an argument could be made that the code is easier to read/maintain if the `Some` is at the use site instead of returned from a function, I'm not sure it's enough to be worth the hassle.

Reply:
Agree, and if it was more generally used I'd replace it, but since it's really only ever going to be called here, I think this is okay. Happy to change it if you feel strongly though :)

Reply:
No need, I was just brainstorming.